### PR TITLE
Search Functionality In Category Screen

### LINKED ISSUE #3488: [FEATURE] Category Search Bar
### Please confirm the following:

- [X] I've checked [the current issues](https://github.com/Ivy-Apps/ivy-wallet/issues) for duplicate issues.
- [X] I've requested a single (only one) feature/change in this issue. It complies with the One Request Per GitHub Issue (ORPGI) rule.
- [X] My issue is well-defined and describes how it should be implemented from UI/UX perspective.

### What do you want to be added or improved?

In Category Screen as a user I want to have a search functionality via search bar
[
category_screen
](url)

### Why do you need it?

- We have good number of categories and user can add custom categories. Therefore the list may get longer. So it'd be best to have a search option , so one can easily find desired category in the screen.

### How do you imagine it?

In category screen we can have a search bar. This bar will be hidden by default. User can make it visible  from settings.

### PR DESCRIPTION
**Pull request by shamim-emon**
## Pull request (PR) checklist

Please check if your pull request fulfills the following requirements:

- [x] I've read the [Contribution Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/CONTRIBUTING.md) and my PR doesn't break the rules.
- [x] I've read and understand the [Developer Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/Guidelines.md).
- [x] I confirm that I've run the code locally and everything works as expected.
- [x] My PR includes only the necessary changes to fix the issue (i.e., no unnecessary files or lines of code are changed).
- [x] 🎬 I've attached a **screen recording** of using the new code to the next paragraph (if applicable).

## What's changed?

Describe with a few bullets **what's new:**
-  I have added optional search functionality in `Category Screen`

https://github.com/user-attachments/assets/6fa8c504-7433-4bcb-96c1-e42685fdaa84

## Does this PR close any GitHub issues? (do not delete)

- Closes #3488 

## Troubleshooting GitHub Actions (CI) failures ❌
Pull request checks failing? Read our [CI Troubleshooting guide](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/CI-Troubleshooting.md).

### CODE REVIEW

File: shared/domain/src/main/java/com/ivy/domain/features/IvyFeatures.kt [hunk 1]
```diff
@@ -32,6 +32,13 @@ class IvyFeatures @Inject constructor() : Features {
         defaultValue = true
     )
 
+    override val showCategorySearchBar = BoolFeature(
+        key = "show_category_search_bar",
+        name = "Show category search bar",
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/features/IvyFeatures.kt:37]: Search categories

File: screen/categories/src/main/java/com/ivy/categories/CategoriesScreen.kt [hunk 2]
```diff
@@ -752,6 +772,106 @@ private fun Preview(theme: Theme = Theme.LIGHT, compactModeEnabled: Boolean = fa
...
 }
 
+@Preview
+@Composable
+private fun PreviewWithSearchBarEnabled(
+    theme: Theme = Theme.LIGHT,
+    compactModeEnabled: Boolean = false,
+    displaySearchBarEnabled: Boolean = true
+) {
+    com.ivy.legacy.IvyWalletPreview(theme) {
```
Comment (ILIYANGERMANOV) [inline screen/categories/src/main/java/com/ivy/categories/CategoriesScreen.kt:782]: Import preview

File: screen/categories/src/main/java/com/ivy/categories/CategoriesViewModel.kt [hunk 3]
```diff
@@ -55,6 +55,8 @@ class CategoriesViewModel @Inject constructor(
     private val baseCurrency = mutableStateOf("")
     private val categories =
         mutableStateOf<ImmutableList<CategoryData>>(persistentListOf<CategoryData>())
+    private val allCategories =
```
Comment (ILIYANGERMANOV) [inline screen/categories/src/main/java/com/ivy/categories/CategoriesViewModel.kt:58]: We don't need this state. Just in the `@Composable getCategories`()` do filter the categories state by the query. You alsoo need to `remember(categories, query) { categories.filter { ... } }`

File: screen/categories/src/main/java/com/ivy/categories/CategoriesViewModel.kt [hunk 4]
```diff
@@ -171,9 +183,21 @@ class CategoriesViewModel @Inject constructor(
             }
 
             val sortedList = sortList(categories, sortOrder.value).toImmutableList()
+            this.allCategories.value = sortedList
+            this.categories.value = this.allCategories.value
+        }
+    }
 
-            this.categories.value = sortedList
+    private fun filterCategories(queryString: String) {
```
Comment (ILIYANGERMANOV) [inline screen/categories/src/main/java/com/ivy/categories/CategoriesViewModel.kt:191]: No need for this, in pseudo code it should look like this

```kotlin
@Composable
fun getCategories() {
  val allCats = categories.value
  return remember(allCats, query.value) {
    allCats.filter { ... }.toImmutableList()
  }
}
```

File: screen/categories/src/main/java/com/ivy/categories/CategoriesViewModel.kt [hunk 5]
```diff
@@ -171,9 +183,21 @@ class CategoriesViewModel @Inject constructor(
...
+        var unsortedList: List<CategoryData>
+        if (queryString.isNotEmpty()) {
+            unsortedList = this.allCategories.value.filter {
+                it.category.name.value.toLowerCase().contains(queryString.toLowerCase().trim())
```
Comment (ILIYANGERMANOV) [inline screen/categories/src/main/java/com/ivy/categories/CategoriesViewModel.kt:195]: It should be:
```kotlin
query.lowercase().trim() in it.category.name.lowercase()
```

Logically, we're looking of the existence of a keyword/query in a category name and not the other way around

**Review by ILIYANGERMANOV — CHANGES_REQUESTED** [review]

File: shared/domain/src/main/java/com/ivy/domain/features/IvyFeatures.kt (hunk 1 above)
Comment (shamim-emon) [reply shared/domain/src/main/java/com/ivy/domain/features/IvyFeatures.kt]: Done

File: screen/categories/src/main/java/com/ivy/categories/CategoriesScreen.kt (hunk 2 above)
Comment (shamim-emon) [reply screen/categories/src/main/java/com/ivy/categories/CategoriesScreen.kt]: Done

File: screen/categories/src/main/java/com/ivy/categories/CategoriesViewModel.kt (hunk 4 above)
Comment (shamim-emon) [reply screen/categories/src/main/java/com/ivy/categories/CategoriesViewModel.kt]: Done

File: screen/categories/src/main/java/com/ivy/categories/CategoriesViewModel.kt (hunk 5 above)
Comment (shamim-emon) [reply screen/categories/src/main/java/com/ivy/categories/CategoriesViewModel.kt]: Done

**Review by ILIYANGERMANOV — APPROVED** [review]
Nicely done 👏

File: shared/domain/src/main/java/com/ivy/domain/features/IvyFeatures.kt [hunk 6]
```diff
@@ -32,6 +32,13 @@ class IvyFeatures @Inject constructor() : Features {
         defaultValue = true
     )
 
+    override val showCategorySearchBar = BoolFeature(
+        key = "search_categories",
+        name = "Search categories",
+        description = "Show search bar in category screen",
+        defaultValue = false
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/features/IvyFeatures.kt:39]: Let's enable this feature by default 🙌

**Comment by ILIYANGERMANOV** [comment]
@shinebrillant great work! As a feedback, try putting meaningful commit message describing the change in the commit briefly

File: shared/domain/src/main/java/com/ivy/domain/features/IvyFeatures.kt (hunk 6 above)
Comment (ILIYANGERMANOV) [reply shared/domain/src/main/java/com/ivy/domain/features/IvyFeatures.kt:39]: @shamim-emon enable it bt default and we can merge 🚀 Ping me when it's ready (CI passed)

Comment (shamim-emon) [reply shared/domain/src/main/java/com/ivy/domain/features/IvyFeatures.kt]: @ILIYANGERMANOV It's  ready!

**Pull request merged**